---
title: "ArchR_tutorial"
author: "Ronald Wu"
date: '2024-05-03'
output: html_document
---

# Single Cell ATAC-seq Analysis
There are a few well known packages in R and python that can be used to analyze scATAC-seq data.
+ R Packages:
  + Signac: developed by Tim Stuart while in the Satija lab. Actively maintained by Stuart lab.
    + [website](https://stuartlab.org/signac/)
  + ArchR: 
    + [website](https://www.archrproject.com/)

# ArchR Vignettes
The purpose of this notebook is to learn how to use `ArchR` for processing and analyzing scATAC-seq data. We'll be following the vignettes provided by the Greenleaf lab, which is the original developer of this package. All code is slightly modified from their original vignettes from the following:
+ [brief tutorial](https://www.archrproject.com/articles/Articles/tutorial.html)
+ 

# Setup
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Install packages

```{r install packages, eval = FALSE}
# install devtools: allows install of packages from github
if (!requireNamespace("devtools", quietly = TRUE)) install.packages("devtools")

# if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

devtools::install_github("GreenleafLab/ArchR", ref="master", repos = BiocManager::repositories())

library(ArchR)
ArchR::installExtraPackages() # installs Cairo and other stuff

 library(BiocManager)
      install("BSgenome.Hsapiens.UCSC.hg19") # needed to set addArchRGenome to hg19

install.packages("hexbin") # for plotting

install.packages("harmony") # for batch correction
```

I also installed `macs2` version 2.2.9.1 locally (on macbook pro) using the command `pip install macs2`. MACS2 will be called by ArchR for peak calling purposes. Note: There is a newer versions of MACS called `MACS3` but I don't believe is it support by ArchR.

# Load Libraries
```{r load libraries}
library(ArchR) # version 1.0.2
```

# 1. Brief tutorial

## Setup and download tutorial data
Below we are following the [brief tutorial](https://www.archrproject.com/articles/Articles/tutorial.html).

```{r Setup and download tutorial data, eval = FALSE}
library(ArchR)
set.seed(1)
addArchRThreads(threads = 2) 

# download tutorial data (0.5 gigs to current working directory)
inputFiles <- getTutorialData("Hematopoiesis")

# change location of tutorial files to data/
temp <- names(inputFiles)
inputFiles <- paste0("data/", inputFiles)
names(inputFiles) <- temp
```
After downloading the tutorial files, we manually moved them to the `data/` folder in the current working directory.

## Creating Arrow Files

```{r creating arrow files}
# set genome to hg19
addArchRGenome("hg19")

# create arrow files (subset, first sample only)
ArrowFiles <- createArrowFiles(
  inputFiles = inputFiles[1],
  sampleNames = names(inputFiles)[1],
  minTSS = 4, # Dont set this too high because you can always increase later
  minFrags = 1000, 
  addTileMat = TRUE,
  addGeneScoreMat = TRUE
)

ArrowFiles

```
Note: we modified the above tutorial code to process only 1 of 3 samples. Had an error when all 3 inputFiles were used.

When only the first sample `scATAC_BMMC_R1` was used, arrow file was created without error. Still took 32 mins to complete.

Two matrices are created: 
1) GeneScoreMatrix: the atac-seq signal proximal to the TSS is used to make an estimate of the gene activity score.
2) TileMatrix: A counts matrix where the genome is tiled into a fixed-width sliding window. Not the usual peak based matrix.

## Inferring doublets

```{r inferring doublets}
doubScores <- addDoubletScores(
  input = ArrowFiles,
  k = 10, #Refers to how many cells near a "pseudo-doublet" to count.
  knnMethod = "UMAP", #Refers to the embedding to use for nearest neighbor search.
  LSIMethod = 1
)

doubScores

```

The doublet results report are found in the `QualityControl/scATAC_BMMC_R1` folder. Note, the doublet scores are added to the arrowfiles.

## Creating an ArchRProject

```{r create ArchRProject}
proj <- ArchRProject(
  ArrowFiles = ArrowFiles, 
  outputDirectory = "HemeTutorial",
  copyArrows = TRUE #This is recommended so that you maintain an unaltered copy for later usage.
)

# show available matrices for the project 
getAvailableMatrices(proj) #  "GeneScoreMatrix" "TileMatrix" 

# filter doublets; note addDoubletScores must be run previously
## Default filterRatio = 1; this is a consistent filter applied on all samples
## Can be adjusted to filter more cells
proj <- filterDoublets(ArchRProj = proj) 
```

An ArchRProject must be created for further analysis of our arrowfiles. About 4.9% cells were doublets and filtered out. Thus, 4689 cells were retained.

## Dimensionality Reduction and Clustering

```{r dimensionality reduction and cluster}
# IterativeLSI used for dimensionality reduction
proj <- addIterativeLSI(ArchRProj = proj, useMatrix = "TileMatrix", name = "IterativeLSI")

# runs seurat's FindClusters function
proj <- addClusters(input = proj, 
                    reducedDims = "IterativeLSI" # name of the reducedDims object
                    )


```

IterativeLSI is used for dimensionality reduction. By default, the log(tf - idf) method is used. Can also be changed to tf - log(idf) or log(tf) - log(idf). Importantly, iterativeLSI is not deterministic and results can slightly differ if run another time on the same data.

There are 8 Clusters identified. Cluster identification is deterministic. 

## Visualizing in a 2D UMAP Embedding

```{r UMAP Visualization}
# add UMAP embedding
proj <- addUMAP(ArchRProj = proj, reducedDims = "IterativeLSI")

# plot UMAP:
## colorBy Sample
p1 <- plotEmbedding(ArchRProj = proj, colorBy = "cellColData", name = "Sample", embedding = "UMAP")
## colorBy Clusters
p2 <- plotEmbedding(ArchRProj = proj, colorBy = "cellColData", name = "Clusters", embedding = "UMAP")

p1 + p2

# ggAlignPlots(p1, p2, type = "h")
```
### save plots to pdf
```{r save plots to pdf, eval = FALSE}
plotPDF(p1,p2, # vector of plots to plot
        name = "2024-05-06-BMMC-Plot-UMAP-Sample-Clusters.pdf", # file name
        ArchRProj = proj, 
        addDOC = FALSE, width = 5, height = 5)
```

## Assigning Clusters with Gene Scores

Given a vector of marker genes, we can overlay the gene scores of those markers onto the UMAP. MAGIC is used for marker gene imputation, this allows for a better visualization of the data, since sparcity may make plots look quite variable. 

```{r cluster assignment with gene scores}

# Marker gene imputation with MAGIC
proj <- addImputeWeights(proj)

# marker genes to use to assign clusters
markerGenes  <- c(
    "CD34",  #Early Progenitor
    "GATA1", #Erythroid
    "PAX5", "MS4A1", "MME", #B-Cell Trajectory
    "CD14", "MPO", #Monocytes
    "CD3D", "CD8A"#TCells
  )
```

```{r plot umap with marker genes}
# plot UMAP 
## colorBy markerGenes 
p <- plotEmbedding(
    ArchRProj = proj, 
    colorBy = "GeneScoreMatrix", 
    name = markerGenes, 
    embedding = "UMAP",
    imputeWeights = getImputeWeights(proj)
)

# plot a specific marker
p$CD14

#Rearrange for grid plotting
p3 <- lapply(p, function(x){
    x + guides(color = FALSE, fill = FALSE) + 
    theme_ArchR(baseSize = 6.5) +
    theme(plot.margin = unit(c(0, 0, 0, 0), "cm")) +
    theme(
        axis.text.x=element_blank(), 
        axis.ticks.x=element_blank(), 
        axis.text.y=element_blank(), 
        axis.ticks.y=element_blank()
    )
})
do.call(cowplot::plot_grid, c(list(ncol = 3),p3))
```

The MAGIC diffusion matrix is used smooth dropout noise in the gene score matrix. 

### save vector of plots to pdf

```{r save vector of plots to pdf}
plotPDF(plotList = p, 
    name = "2024-05-06-BMMC-Plot-UMAP-Marker-Genes-W-Imputation.pdf", 
    ArchRProj = proj, # plots are saved into Plots/ in the assigned outputDirectory
    addDOC = FALSE, width = 5, height = 5)
```

## Visualizing Genome Browser Tracks
```{r Visualizing Genome Browser Tracks}

# visualize the genome by the marker genes and by clusters
p4 <- plotBrowserTrack(
    ArchRProj = proj, 
    groupBy = "Clusters", 
    geneSymbol = markerGenes, 
    upstream = 50000,
    downstream = 50000
)

# plot a specific gene
grid::grid.newpage()
grid::grid.draw(p4$CD14)
grid::grid.draw(p4$GATA1)
```
### save PDF of genome browser tracks
```{r pdf of genome browser tracks}

plotPDF(plotList = p4, 
    name = "2024-05-06-Plot-Tracks-Marker-Genes.pdf", 
    ArchRProj = proj, 
    addDOC = FALSE, width = 5, height = 5)
```

## Saving and Loading an ArchRProject
```{r saving and loading project}
# save
# proj <- saveArchRProject(ArchRProj = proj)

# load
# proj <- loadArchRProject(path = "HemeTutorial")
```

By default, saving the project will overwrite the files in the output directory `HemeTutorial/`, which we assigned earlier when creating the ArchRProject object.

## Session info
```{r session info}
Sys.Date()
sessionInfo()
```

# Advanced tutorial

The advanced tutorial builds upon the foundation from the basic tutorial, which covered the typical workflow for preparing the ArchRProject object, dimensionality reduction, clustering and plotting. 

## Creating Arrow Files

Here we create the arrow files from the all three tutorial datasets found in the `HemeFragments/` folder. Previously, we used only one sample due to an error with using all three at once. The error was due to project path having a space in it. This is a common issue with dropbox folders.

```{r creating arrow files, eval = FALSE }
# set genome to hg19
addArchRGenome("hg19")

# create arrow files (subset, first sample only)
ArrowFiles_all <- createArrowFiles(
  inputFiles = inputFiles,
  sampleNames = names(inputFiles),
  minTSS = 4, # Dont set this too high because you can always increase later
  minFrags = 1000, 
  addTileMat = TRUE,
  addGeneScoreMat = TRUE
)

ArrowFiles_all

```

## Create ArchRProject
```{r create ArchRProject}
projHeme1 <- ArchRProject(
  ArrowFiles = ArrowFiles_all, 
  outputDirectory = "HemeTutorial",
  copyArrows = TRUE #This is recommended so that you maintain an unaltered copy for later usage.
)

# show available matrices for the project 
getAvailableMatrices(projHeme1) #  "GeneScoreMatrix" "TileMatrix" 

# filter doublets; note addDoubletScores must be run previously
## Default filterRatio = 1; this is a consistent filter applied on all samples
## Can be adjusted to filter more cells
projHeme1 <- filterDoublets(ArchRProj = projHeme1) 
```

## Accessing the ArchRProject object
```{r object access}

# accessing the cellColData matrix in the ArchRProject object
projHeme1$cellNames
projHeme1$Sample
projHeme1$TSSEnrichment

quantile(projHeme1$TSSEnrichment)

# Subset options:
## 1. Take first 100 cells

projHeme1[1:100,]
projHeme1$cellNames[1:100]

# 2. keep specific sample "scATAC_BMMC_R1"
idxSample <- BiocGenerics::which(projHeme1$Sample %in% "scATAC_BMMC_R1")
cellsSample <- projHeme1$cellNames[idxSample]
projHeme1[cellsSample, ]

# 3. keep cells based on TSS enrichment score (previously minTSS = 4)
quantile(projHeme1$TSSEnrichment)
hist(projHeme1$TSSEnrichment)

idxPass <- which(projHeme1$TSSEnrichment >= 8)
cellsPass <- projHeme1$cellNames[idxPass]
projHeme1[cellsPass, ]

# 4. Adding data to ArchRProject object

## shortening sample names
projHeme1$Sample %>% factor
bioNames <- gsub("_R2|_R1|scATAC_","",projHeme1$Sample)
head(bioNames)

projHeme1$bioNames <- bioNames # add bioNames column
  # addCellColData() function can also be used

# 5. Accessing and operating on a specific column
getCellColData(projHeme1)
df <- getCellColData(projHeme1, select = "nFrags")
df
df <- getCellColData(projHeme1, select = c("log10(nFrags)", "nFrags - 1")) # subtraction done here
df
```

# Plot Sample QC statistics

```{r plot sample QC stats}

# 1. Ridge plot of TSS Enrichment score
p5 <- plotGroups(
  ArchRProj = projHeme1, 
    groupBy = "Sample", 
    colorBy = "cellColData", 
    name = "TSSEnrichment",
    plotAs = "ridges"
)

# 2. Violin plot of TSS Enrichment score
p6 <- plotGroups(
    ArchRProj = projHeme1, 
    groupBy = "Sample", 
    colorBy = "cellColData", 
    name = "TSSEnrichment",
    plotAs = "violin",
    alpha = 0.4,
    addBoxPlot = TRUE
   )

# 3. Ridge plot for log10(unique nuclear fragments)
p7 <- plotGroups(
  ArchRProj = projHeme1, 
    groupBy = "Sample", 
    colorBy = "cellColData", 
    name = "log10(nFrags)",
    plotAs = "ridges"
)

# 4. Violin plot for log10(unique nuclear fragments)
p8 <- plotGroups(
  ArchRProj = projHeme1, 
    groupBy = "Sample", 
    colorBy = "cellColData", 
    name = "log10(nFrags)",
    plotAs = "violin",
    alpha = 0.4,
    addBoxPlot = TRUE
)

p5 + 
  ggtitle("TSS Enrichment Score - Ridge Plot") +
  scale_fill_brewer(palette = "Dark2")

p6 + 
  ggtitle("TSS Enrichment Score - Violin Plot") +
  scale_fill_brewer(palette = "Dark2")

p7 + 
  ggtitle("log10(unique nuclear fragments) - Ridge Plot") +
  scale_fill_brewer(palette = "Dark2")

p8 + 
  ggtitle("log10(unique nuclear fragments) - Violin Plot") +
  scale_fill_brewer(palette = "Dark2")
```
Note: These plots show the TSSEnrichment scores after filtering via `minTSS = 4` as selected using the `createArrowFiles()`.

## save pdf of QC stats
```{r save pdf of QC states, eval = FALSE}
plotPDF(p5,p6,p7,p8, 
        name = "QC-Sample-Statistics.pdf", 
        ArchRProj = projHeme1, 
        addDOC = TRUE, # appends date to end of filename
        width = 4, 
        height = 4
        )

# plot list (additional ggplot layers are added here)

plotList = list(a = p5 + 
                  ggtitle("TSS Enrichment Score - Ridge Plot") +
                  scale_fill_brewer(palette = "Dark2"), 
                b = p6 + 
                  ggtitle("TSS Enrichment Score - Violin Plot") +
                  scale_fill_brewer(palette = "Dark2"),
                c = p7 + 
                  ggtitle("log10(unique nuclear fragments) - Ridge Plot") +
                  scale_fill_brewer(palette = "Dark2"),
                d = p8 + 
                  ggtitle("log10(unique nuclear fragments) - Violin Plot") +
                  scale_fill_brewer(palette = "Dark2")
                )

plotPDF(plotList = plotList,
        name = "plotList-QC-Sample-Statistics.pdf", 
        ArchRProj = projHeme1, 
        addDOC = TRUE, 
        width = 4, 
        height = 4
        )
```
# remove doublets
```{r remove doublets, eval = FALSE}
# for all 3 samples

doubScores_all <- addDoubletScores(
  input = ArrowFiles_all,
  k = 10, #Refers to how many cells near a "pseudo-doublet" to count.
  knnMethod = "UMAP", #Refers to the embedding to use for nearest neighbor search.
  LSIMethod = 1
)

doubScores_all
```

# Additional Batch Correction with Harmony 
```{r harmony batch correction}
projHeme2 <- projHeme1

# dimensionality reduction
projHeme2 <- addIterativeLSI(
    ArchRProj = projHeme2,
    useMatrix = "TileMatrix", 
    name = "IterativeLSI", 
    iterations = 2, 
    clusterParams = list( #See Seurat::FindClusters
        resolution = c(0.2), 
        sampleCells = 10000, 
        n.start = 10
    ), 
    varFeatures = 25000, 
    dimsToUse = 1:30
)

# run harmony batch correction
projHeme2 <- addHarmony(
    ArchRProj = projHeme2,
    reducedDims = "IterativeLSI",
    name = "Harmony",
    groupBy = "Sample"
)

# new reducedDims object is created
projHeme2@reducedDims$Harmony
projHeme2@reducedDims$IterativeLSI # the IterativeLSI created previously


```

# Clustering

```{r clustering}

# add clusters based on IterativeLSI
projHeme2 <- addClusters(
    input = projHeme2,
    reducedDims = "IterativeLSI",
    method = "Seurat",
    name = "Clusters",
    resolution = 0.8
)

# add clusters based on HArmony batch correction
projHeme2 <- addClusters(
    input = projHeme2,
    reducedDims = "Harmony",
    method = "Seurat",
    name = "Clusters_harmony",
    resolution = 0.8
)

# check clusters

projHeme2$Clusters %>% table
projHeme2$Clusters_harmony %>% table
```

IterativeLSI followed by clustering shows 12 clusters, whereas harmony has 13 total.

## Heatmap of cluster membership of samples
```{r confusion matrix and heatmap }
library(pheatmap)

# 1. IterativeLSI clusters
# make confusion matrix
cM <- confusionMatrix(paste0(projHeme2$Clusters), paste0(projHeme2$Sample))
cM

# normalize
cM <- cM / Matrix::rowSums(cM)

# heatmap
p <- pheatmap::pheatmap(
    mat = as.matrix(cM), 
    color = paletteContinuous("whiteBlue"), 
    border_color = "black"
)
p

# 2. Harmony clusters

# make confusion matrix
cM_2 <- confusionMatrix(paste0(projHeme2$Clusters_harmony), paste0(projHeme2$Sample))
cM_2

# normalize
cM_2 <- cM_2 / Matrix::rowSums(cM_2)

# heatmap
p2 <- pheatmap::pheatmap(
    mat = as.matrix(cM_2), 
    color = paletteContinuous("whiteBlue"), 
    border_color = "black"
)
p2

p
p2
```


# UMAP

```{r UMAP}
projHeme2 <- addUMAP(
    ArchRProj = projHeme2, 
    reducedDims = "IterativeLSI", 
    name = "UMAP", 
    nNeighbors = 30, 
    minDist = 0.5, 
    metric = "cosine"
)

# adds harmony to embeddings
projHeme2 <- addUMAP(
    ArchRProj = projHeme2, 
    reducedDims = "Harmony", 
    name = "UMAP_harmony", 
    nNeighbors = 30, 
    minDist = 0.5, 
    metric = "cosine"
)

# access UMAP embeddings

projHeme2@embeddings$UMAP
projHeme2@embeddings$UMAP_harmony %>% View
```

## Plot umap
```{r plot umap}

# plot by sample
p1 <- plotEmbedding(ArchRProj = projHeme2, 
                    colorBy = "cellColData", 
                    name = "Sample", 
                    embedding = "UMAP"
                    )


p2 <- plotEmbedding(ArchRProj = projHeme2, 
                    colorBy = "cellColData", 
                    name = "Sample", 
                    embedding = "UMAP_harmony"
                    )
ggAlignPlots(p1, p2, type = "h")

# plot by clusters
p3 <- plotEmbedding(ArchRProj = projHeme2, 
                    colorBy = "cellColData", 
                    name = "Clusters", 
                    embedding = "UMAP"
                    )

p4 <- plotEmbedding(ArchRProj = projHeme2, 
                    colorBy = "cellColData", 
                    name = "Clusters_harmony", 
                    embedding = "UMAP_harmony"
                    )


ggAlignPlots(p3, p4, type = "h")


# plot the LSI embedding with harmony labels
p5 <- plotEmbedding(ArchRProj = projHeme2, 
                    colorBy = "cellColData", 
                    name = "Clusters_harmony", 
                    embedding = "UMAP"
                    )
ggAlignPlots(p3, p4, p5, type = "h")


```
## save PDF of UMAPs
```{r save pdf of umaps, eval = FALSE}

plotPDF(p1,p2,p3,p4,p5,
        name = "Plot-UMAP-Sample-HarmonyClusters.pdf", 
        ArchRProj = projHeme2, 
        addDOC = TRUE, 
        width = 5, height = 5)

```

## Saving and Loading an ArchRProject
```{r saving and loading project, eval = FALSE}
# check where output directory is assigned
getOutputDirectory(projHeme2) # HemeTutorial/

# save
projHeme2 <- saveArchRProject(ArchRProj = projHeme2)

# load
# proj <- loadArchRProject(path = "HemeTutorial")
```

By default, saving the project will overwrite the files in the output directory `HemeTutorial/`, which we assigned earlier when creating the ArchRProject object.