---
title: "ArchR_Terekhanova_hypoxia"
author: "Ronald Wu"
date: '2024-06-18'
output: html_document
---

# Introduction

# Setup
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Install packages

```{r install packages, eval = FALSE}
# install devtools: allows install of packages from github
if (!requireNamespace("devtools", quietly = TRUE)) install.packages("devtools")

# if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

devtools::install_github("GreenleafLab/ArchR", ref="master", repos = BiocManager::repositories())

library(ArchR)
ArchR::installExtraPackages() # installs Cairo and other stuff

 library(BiocManager)
      install("BSgenome.Hsapiens.UCSC.hg19") # needed to set addArchRGenome to hg19

install.packages("hexbin") # for plotting

install.packages("harmony") # for batch correction
```

# Load Libraries
```{r load libraries}
library(ArchR) # version 1.0.2
library("here")
library("BSgenome.Hsapiens.UCSC.hg38")
```

# Download Terekhanova GBM Samples

The Terekhanova samples are available on [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE240822). There are many different tumour types here and also snRNA data. We are interested in the 18 snATAC GBM samples.

Here we use the command line to download the files, extract the archive and delete the archives. 
```{bash download samples, eval = FALSE}

# Make a folder to store the downloaded files
dir="data/Terekhanova"

if [ ! -d "$dir" ]; then
  echo "Creating $dir directory..."
  mkdir -p "$dir"
else
  echo "$dir directory already exists."
fi

# change directory
cd data/Terekhanova/

# download the first 4
wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710021/suppl/GSM7710021_C3L-02705_CPT0189650015_snATAC_GBM.tar.gz

wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710022/suppl/GSM7710022_C3L-03405_CPT0224600013_snATAC_GBM.tar.gz

wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710023/suppl/GSM7710023_C3L-03968_CPT0228220004_snATAC_GBM.tar.gz

wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710024/suppl/GSM7710024_C3N-00662_CPT0087680014_snATAC_GBM.tar.gz

#
wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710025/suppl/GSM7710025_C3N-00663_CPT0087730014_snATAC_GBM.tar.gz

wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710026/suppl/GSM7710026_C3N-01334_CPT0125220004_snATAC_GBM.tar.gz

wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710027/suppl/GSM7710027_C3N-01518_CPT0167640014_snATAC_GBM.tar.gz

wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710028/suppl/GSM7710028_C3N-01798_CPT0167750015_snATAC_GBM.tar.gz

wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710029/suppl/GSM7710029_C3N-01814_CPT0167860015_snATAC_GBM.tar.gz

#
wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710030/suppl/GSM7710030_C3N-01816_CPT0168080014_snATAC_GBM.tar.gz

wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710031/suppl/GSM7710031_C3N-01818_CPT0168270014_snATAC_GBM.tar.gz

wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710032/suppl/GSM7710032_C3N-02181_CPT0168380014_snATAC_GBM.tar.gz

wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710033/suppl/GSM7710033_C3N-02186_CPT0168720014_snATAC_GBM.tar.gz

wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710034/suppl/GSM7710034_C3N-02188_CPT0168830014_snATAC_GBM.tar.gz

wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710035/suppl/GSM7710035_C3N-02769_CPT0205570014_snATAC_GBM.tar.gz

wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710036/suppl/GSM7710036_C3N-02783_CPT0205890014_snATAC_GBM.tar.gz

wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710037/suppl/GSM7710037_C3N-02784_CPT0206000015_snATAC_GBM.tar.gz

wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM7710nnn/GSM7710038/suppl/GSM7710038_C3N-03186_CPT0206880004_snATAC_GBM.tar.gz

# extract archives and delete archive

for file in *.tar.gz; do tar -xzf "$file" && rm "$file"; done

```

# ArchR Setup

```{r ArchR Setup}
set.seed(1)
addArchRThreads(threads = 2) 

parent_dir <- here("data","Terekhanova")

# Get a list of subdirectories 
subdirs <- list.dirs(parent_dir, recursive = FALSE, full.names = FALSE)
# Extract prefixes before the underscore to get sample names
prefixes <- sapply(subdirs, function(x) strsplit(x, "_")[[1]][1])

# Get list of relative path to fragments.tsv.gz for all Terekhanova samples
list_files <- function(dir_path) {
  full_paths <- list.files(dir_path, recursive = TRUE, full.names = FALSE, pattern = "fragments.tsv.gz$")
  return(full_paths)
}
matching_files <- list_files(dir_path = here("data","Terekhanova"))

# make character vector with sample name and full path to their respective fragments.tsv.gz
inputFiles1 <- paste0(here("data","Terekhanova",matching_files))
names(inputFiles1) <- prefixes

# check inputFiles1
inputFiles1
```

# Creating Arrow Files

```{r creating arrow files}
# set genome to hg38
addArchRGenome("hg38")

# create arrow files
ArrowFiles <- createArrowFiles(
  inputFiles = inputFiles1,
  sampleNames = names(inputFiles1),
  minTSS = 4, # Dont set this too high because you can always increase later
  minFrags = 1000, 
  addTileMat = TRUE,
  addGeneScoreMat = TRUE
)

ArrowFiles

# create single arrow file for C3L-03968 alone :
# ArrowFiles1 <- createArrowFiles(
#   inputFiles = inputFiles1[3],
#   sampleNames = names(inputFiles1)[3],
#   minTSS = 4, # Dont set this too high because you can always increase later
#   minFrags = 1000, 
#   addTileMat = TRUE,
#   addGeneScoreMat = TRUE
# )

# consolidate all the arrowfiles names
#ArrowFiles <- c(ArrowFiles,ArrowFiles1)
```


## Inferring doublets

```{r inferring doublets}
doubScores <- addDoubletScores(
  input = ArrowFiles,
  k = 10, #Refers to how many cells near a "pseudo-doublet" to count.
  knnMethod = "UMAP", #Refers to the embedding to use for nearest neighbor search.
  LSIMethod = 1
)

doubScores

```

# Create ArchRProject

```{r create ArchRProject, eval = FALSE}
proj <- ArchRProject(
  ArrowFiles = ArrowFiles, 
  outputDirectory = "Terekhanova",
  copyArrows = FALSE #This is recommended so that you maintain an unaltered copy for later usage.
)

# show available matrices for the project 
getAvailableMatrices(proj) #  "GeneScoreMatrix" "TileMatrix" 

# filter doublets; note addDoubletScores must be run previously
## Default filterRatio = 1; this is a consistent filter applied on all samples
## Can be adjusted to filter more cells
proj <- filterDoublets(ArchRProj = proj) 
```
	Filtering 2755 cells from ArchRProject!
	C3N-00662 : 629 of 7937 (7.9%)
	C3L-03405 : 987 of 9939 (9.9%)
	C3L-02705 : 646 of 8040 (8%)
	C3L-03968 : 493 of 7026 (7%)
	
# Plot Sample QC statistics

```{r plot sample QC stats}

# Fragment Size Distribution

p1 <- plotFragmentSizes(ArchRProj = proj)

# TSS Enrichment profile
p2 <- plotTSSEnrichment(ArchRProj = proj)


# 1. Ridge plot of TSS Enrichment score
p5 <- plotGroups(
  ArchRProj = proj, 
    groupBy = "Sample", 
    colorBy = "cellColData", 
    name = "TSSEnrichment",
    plotAs = "ridges"
)

# 2. Violin plot of TSS Enrichment score
p6 <- plotGroups(
    ArchRProj = proj, 
    groupBy = "Sample", 
    colorBy = "cellColData", 
    name = "TSSEnrichment",
    plotAs = "violin",
    alpha = 0.4,
    addBoxPlot = TRUE
   )

# 3. Ridge plot for log10(unique nuclear fragments)
p7 <- plotGroups(
  ArchRProj = proj, 
    groupBy = "Sample", 
    colorBy = "cellColData", 
    name = "log10(nFrags)",
    plotAs = "ridges"
)

# 4. Violin plot for log10(unique nuclear fragments)
p8 <- plotGroups(
  ArchRProj = proj, 
    groupBy = "Sample", 
    colorBy = "cellColData", 
    name = "log10(nFrags)",
    plotAs = "violin",
    alpha = 0.4,
    addBoxPlot = TRUE
)

p1 + 
  ggtitle("Fragment Sizes") +
  scale_fill_brewer(palette = "Dark2")

p2 + 
  ggtitle("TSS Enrichment Profile") +
  scale_fill_brewer(palette = "Dark2")

p5 + 
  ggtitle("TSS Enrichment Score - Ridge Plot") +
  scale_fill_brewer(palette = "Dark2")

p6 + 
  ggtitle("TSS Enrichment Score - Violin Plot") +
  scale_fill_brewer(palette = "Dark2")

p7 + 
  ggtitle("log10(unique nuclear fragments) - Ridge Plot") +
  scale_fill_brewer(palette = "Dark2")

p8 +
  ggtitle("log10(unique nuclear fragments) - Violin Plot") +
  scale_fill_brewer(palette = "Dark2")
```


# Dimensionality Reduction and Clustering

```{r dimensionality reduction and clustering}
# IterativeLSI used for dimensionality reduction
proj <- addIterativeLSI(ArchRProj = proj, useMatrix = "TileMatrix", name = "IterativeLSI")

# runs seurat's FindClusters function
proj <- addClusters(input = proj, 
                    reducedDims = "IterativeLSI" # name of the reducedDims object
                    )


# table showing # of cells per cluster
proj$Clusters %>% table

# total # of clusters
proj$Clusters %>% table %>% length # 16
```

Here we reduce dimensions using the recommended IterativeLSI method and then run identify clusters using `seurat`'s addCluster() function on these reducedDims. 

# Batch Correction and Clustering

Since there are many individual patient samples in the Terekhanova dataset, we anticipate that this might be an issue that requires batch correction with harmony. Here we use Harmony to integrate the scATAC data for batch correction. Harmony typically operates on reduced dimensions such as those identified in PCA in typical scRNA-seq pipelines. Here in scATAC-seq data, harmony is run on reduced dimensions from IterativeLSI.

```{r batch correction and clustering}
# run harmony batch correction 
proj <- addHarmony(
    ArchRProj = proj,
    reducedDims = "IterativeLSI",
    name = "Harmony",
    groupBy = "Sample" # the variable to correct for
)

# add clusters based on Harmony batch correction
proj <- addClusters(
    input = proj,
    reducedDims = "Harmony",
    method = "Seurat",
    name = "Clusters_harmony",
    resolution = 0.8
)

# table of # of cells per cluster

proj$Clusters_harmony %>% table # harmony clusters

# total harmony clusters
proj$Clusters_harmony %>% table %>% length # 23

```
Here we see an increased number of cells identified after harmony correction (23) compared to the default IterativeLSI (16).

# UMAP
Here we add a UMAP embedding based on the reducedDims object and assign it to the ArchRProject object. Both the IterativeLSI and the Harmony reducedDims are used, so we can compare later for evidence of batch effect correction.

Importantly, this step is for visualization purposes and cluster determination is done prior to UMAP in the LSI sub-space.
```{r UMAP}

# UMAP embeddings with IterativeLSI
proj <- addUMAP(
    ArchRProj = proj, 
    reducedDims = "IterativeLSI", 
    name = "UMAP", 
    nNeighbors = 30, 
    minDist = 0.5, 
    metric = "cosine"
)

# adds harmony to embeddings
proj <- addUMAP(
    ArchRProj = proj, 
    reducedDims = "Harmony", 
    name = "UMAP_harmony", 
    nNeighbors = 30, 
    minDist = 0.5, 
    metric = "cosine"
)

# access UMAP embedding

proj@embeddings$UMAP
proj@embeddings$UMAP_harmony %>% View
```

## Plot umap
```{r plot umap}

# plot by sample (IterativeLSI vs Harmony embeddings)
u1 <- plotEmbedding(ArchRProj = proj, 
                    colorBy = "cellColData", 
                    name = "Sample", 
                    embedding = "UMAP"
                    )

u2 <- plotEmbedding(ArchRProj = proj, 
                    colorBy = "cellColData", 
                    name = "Sample", 
                    embedding = "UMAP_harmony"
                    )

# plot by clusters
u3 <- plotEmbedding(ArchRProj = proj, 
                    colorBy = "cellColData", 
                    name = "Clusters", 
                    embedding = "UMAP"
                    )

u4 <- plotEmbedding(ArchRProj = proj, 
                    colorBy = "cellColData", 
                    name = "Clusters_harmony", 
                    embedding = "UMAP_harmony"
                    )

# plot the LSI embedding with harmony labels
u5 <- plotEmbedding(ArchRProj = proj, 
                    colorBy = "cellColData", 
                    name = "Clusters_harmony", 
                    embedding = "UMAP"
                    )

ggAlignPlots(u1, u2, type = "h")
ggAlignPlots(u3, u4, type = "h")
ggAlignPlots(u3, u4, u5, type = "h")
```
## save PDF of UMAPs
```{r save pdf of umaps, eval = FALSE}

# plot PDFs
## u1 vs u2: plot by sample (IterativeLSI vs Harmony embeddings)
## u3 vs u4: plot by clusters 
plotPDF(u1,u2,u3,u4,u5,
        name = "Terekhanova-Plot-UMAP-Sample-HarmonyClusters.pdf", 
        ArchRProj = proj, 
        addDOC = TRUE, 
        width = 5, height = 5)

```

# Using Gene Scores

## Identifying marker genes for each cluster
Here we set the groupBy parameter to either clusters defined by IterativeLSI or Harmony. The bias parameter in the getMarkerFeatures() function allows us the quantile normalize based on various features, such as TSS and log10(nfrag). This step is crucial for the identification of background cells that each cluster or group can be compared to. Note, ArchR identifies a set of background cells for each group/cluster.

This will return a SummarizedExperiment object containing a list of DataFrame objects for each cluster. 

Took about ~15 mins for each run of getMarkerFeatures(). getMarkerFeatures returns a SummarizedExperiment object. 
```{r marker gene identification}

# using IterativeLSI clusters
markerGenes <- getMarkerFeatures(
  ArchRProj = proj,
  useMatrix = "GeneScoreMatrix",
  groupBy = "Clusters",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  testMethod = "wilcoxon"
)

# using Harmony clusters
markerGene_harmony <- getMarkerFeatures(
  ArchRProj = proj,
  useMatrix = "GeneScoreMatrix",
  groupBy = "Clusters_harmony",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  testMethod = "wilcoxon"
)

```

## Retrieve cluster specific marker genes
Here we can access the marker genes that are stored in as the `SummarizedExperiment` objects using the `getMarkers()` function. Notably, this getMarkers function has a default cutOff "FDR <= 0.1 & Log2FC >= 0.5".

```{r get markers}

markerList <- getMarkers(markerGenes, cutOff = "FDR <= 0.01 & Log2FC >= 1.25")
markerList_harmony <- getMarkers(markerGene_harmony, cutOff = "FDR <= 0.01 & Log2FC >= 1.25")

# look at specific cluster: 6
markerList$C6
markerList_harmony$C6

# set gene of interests to label on heatmap
goi_labels <- c("SOX2", 
                "GFAP", "AQP4", 
                "MBP",
                "CD34"
                )

# heatmap with GOIs

heatmapGS <- plotMarkerHeatmap(
  seMarker = markerGenes,
  cutOff = "FDR <= 0.01 & Log2FC >= 1.25",
  labelMarkers = goi_labels,
  transpose = TRUE
)

# heatmap with top 15 markers for each cluster labelled
heatmapGS2 <- plotMarkerHeatmap(
  seMarker = markerGenes,
  cutOff = "FDR <= 0.01 & Log2FC >= 1.25",
  nlabel = 15,
  transpose = TRUE
)

plotMarkerHeatmap(
  seMarker = markerGenes,
  cutOff = "FDR <= 0.01 & Log2FC >= 1.25",
  nLabel = 5,
  transpose = TRUE
)

# heatmap using the harmony clusters

heatmapGS3 <- plotMarkerHeatmap(
  seMarker = markerGene_harmony,
  cutOff = "FDR <= 0.01 & Log2FC >= 1.25",
  labelMarkers = goi_labels,
  transpose = TRUE
)

# heatmap with top 15 markers for each cluster labelled
heatmapGS4 <- plotMarkerHeatmap(
  seMarker = markerGene_harmony,
  cutOff = "FDR <= 0.01 & Log2FC >= 1.25",
  nLabel = 15,
  transpose = TRUE
)



# draw above heatmaps
ComplexHeatmap::draw(heatmapGS, 
                     heatmap_legend_side = "bot",
                     annotation_legend_side = "bot")

ComplexHeatmap::draw(heatmapGS2, 
                     heatmap_legend_side = "bot",
                     annotation_legend_side = "bot")

# volcano plots
## # are there negative log2FC values?

# function to plot volcano plot given a Summarized Experiment object
plot_volcano <- function(df) {
  markerList <- getMarkers(df,
                           cutOff = "FDR <= 0.1" # remove default cutoff to show all points
                           ) # retrieve list of DataFrame from SummarizedExperiment
  cluster_names <- names(markerList)
  # 
  # z <- lapply(markerList, function(x) {
  #   ggplot(as.data.frame(x), aes(x = Log2FC, y = -log10(FDR))) +
  #     geom_point() +
  #     xlab("Log2FC") +
  #     ylab("-log10(FDR)") +
  #     coord_cartesian(xlim = c(-3, 3)) +
  #     ggtitle(paste0("Volcano Plot: ", cluster_names[i]))
  # })
  # 
  # return(z)
  
  
  z <- lapply(1:length(markerList), function(x){
    ggplot(as.data.frame(markerList[[x]]), aes(x = Log2FC, y = -log10(FDR))) +
      geom_point() +
      xlab("Log2FC") +
      ylab("-log10(FDR)") +
      coord_cartesian(xlim = c(-3, 3)) +
      ggtitle(paste0("Volcano Plot: ", cluster_names[x]))
    
  })
  return(z)
}

# plot volcano plot for all the clusters
plot_volcano(markerGenes)

```

About 2900 markers were identified. These markers were identified using the following cutoff:
"FDR <= 0.01 & Log2FC >= 1.25". Thus, only upregulated genes are identified. To confirm this, I set the cutOff param to not exclude by direction of fold change. The volcano plot for each of the 16 clusters show that indeed there are significant marker genes that are downregulated.



## Identifying cell groups based on gene score signatures and marker genes
During the creation of the arrow file, the gene scores are computed and stored as a GeneScoreMatrix by default. Below, I identify the hypoxic cells using single markers and a hypoxic signature (reference??).

```{r hypoxic gene scores}

# define hypoxic related signature
hypoxicGenes <- c(
  "CA9",
  "SLC2A1",
  "HK2",
  "VEGFA"
)

# delete this later (pulled from AUCell)
hypoxicGenes <- geneSets$DAEG_UP_24

# UMAP with overlay with hypoxicGenes
p <- plotEmbedding(
    ArchRProj = proj, 
    colorBy = "GeneScoreMatrix", 
    name = hypoxicGenes, 
    embedding = "UMAP",
    quantCut = c(0.01, 0.95),
    imputeWeights = NULL
)

p2 <- plotEmbedding(
    ArchRProj = proj, 
    colorBy = "GeneScoreMatrix", 
    name = hypoxicGenes, 
    embedding = "UMAP_harmony",
    quantCut = c(0.01, 0.95),
    imputeWeights = NULL
)

# draw umaps
p
p2

# draw smoothened UMAPS:
  # smooth gene scores by imputing weights using Magic
proj <- addImputeWeights(proj)

p3 <- plotEmbedding(
    ArchRProj = proj, 
    colorBy = "GeneScoreMatrix", 
    name = hypoxicGenes, 
    embedding = "UMAP",
    quantCut = c(0.01, 0.95),
    imputeWeights = getImputeWeights(proj)
)

p4 <- plotEmbedding(
    ArchRProj = proj, 
    colorBy = "GeneScoreMatrix", 
    name = hypoxicGenes, 
    embedding = "UMAP_harmony",
    quantCut = c(0.01, 0.95),
    imputeWeights = getImputeWeights(proj)
)

p3
p4

p2
p4
```
Note: ImputWeights also creates a large file on disk (1.22 GB x 2)

## save PDF of UMAPs
```{r save pdf of umaps with gene scores, eval = FALSE}

# plot PDFs
 
plotPDF(p2,
        name = "Terekhanova-Plot-UMAP-HarmonyClusters-genescore-DAEG_24_UP.pdf", 
        ArchRProj = proj, 
        addDOC = TRUE, 
        width = 5, height = 5)

plotPDF(p4,
        name = "Terekhanova-Plot-UMAP-HarmonyClusters-genescore-imputed-DAEG_24_UP.pdf", 
        ArchRProj = proj, 
        addDOC = TRUE, 
        width = 5, height = 5)

```

Next, cell groups are created based on the above scores. The hypoxic group can be compared to normoxic for further analysis, such as differential gene expression and motif analysis.


```{r gene score matrix, eval = FALSE}

# GeneScoreMatrix access
getAvailableMatrices(proj)

getGenes(proj)

# Access the GeneScoreMatrix as SummarizedExperiment
gs_mat <- getMatrixFromProject(proj,
                     useMatrix = "GeneScoreMatrix")
```
# Using GBmap to label scATAC clusters



# Create Psuedo-bulk Replicates

First, create pseudo-bulk replicates for downstream peak calling via MACS2. Here we define `Clusters_harmony` as the groups from which pseudo-bulk replicates are made from. 

```{r pseudobulk, eval = FALSE}
# Save pseudobulk info as new ArchRProject object
proj$DAEG_UP_24_newThr <- as.factor(proj$DAEG_UP_24_newThr)
proj_hyp <- addGroupCoverages(ArchRProj = proj,
                           groupBy = "DAEG_UP_24_newThr" # group for which the pseudobulk rep is made
                           )

proj$PIMO_pos <- as.factor(proj$PIMO_pos)
proj_hyp <- addGroupCoverages(ArchRProj = proj,
                           groupBy = "PIMO_pos" # group for which the pseudobulk rep is made
                           )

```
The `addGroupCoverages` is a long (13 mins) and memory intensive step that produces large intermediate files in the `Terekhanova/GroupCoverages/DAEG_UP_24_newThr/` folder.
# Calling Peaks

## MACS2 Peak Calling
```{r macs2 peak calling, eval = FALSE}
# PATH to MACS2 (previously installed in macbook)
pathToMacs2 <- findMacs2()

# call peaks
proj_hyp2 <- addReproduciblePeakSet(ArchRProj = proj_hyp,
                       groupBy = "DAEG_UP_24_newThr",
                       pathToMacs2 = pathToMacs2)
```
Took 19 mins to call peaks.

## Explore peaks using GRanges
```{r explore peaks GRanges}
# show peak set in ArchRProject object
proj_hyp2@peakSet

# get the peak set
myPeakSet <- getPeakSet(proj_hyp2)

myPeakSet

# tabulate type of peaks
table(myPeakSet$peakType)


```
There are 4 types of peaks annotated under the peakType column in the `Granges` object:
1. Distal
2. Exonic
3. Intronic
4. Promoter


```{r add peak matrix, eval = FALSE}
# check available matrices
getAvailableMatrices(proj)
getAvailableMatrices(proj_hyp)
getAvailableMatrices(proj_hyp2)
getAvailableMatrices(proj_hyp3)

# add Peak Matrix to new ArchRProject object

proj_hyp3 <- addPeakMatrix(proj_hyp2)

# check available matrices
getAvailableMatrices(proj_hyp3)
```
-took mins

## Marker Peaks
Next, we use the `getMarkerFeatures()` on the PeakMatrix to identify Marker Peaks. This is in contrast to Marker Genes, which we found using the GeneScoreMatrix:

Recall we previously used the `getMarkerFeatures()` function to retrieve a `SummarizedExperiment` object from the GeneScoreMatrix, which is one of the matrices found within the ArchRProject Object.
This was used to identify marker genes associated with each cluster, such as those clusters called by Harmony batch correction.

```{r get marker peaks, eval = FALSE}

# check cluster labels
table(proj_hyp3$Clusters_harmony)

# check hypoxia labels
table(proj_hyp3$DAEG_UP_24_newThr)

# get marker peaks for harmony clusters
# markersPeaks <- getMarkerFeatures(
#   ArchRProj = proj3, 
#   useMatrix = "PeakMatrix",
#   groupBy = "Clusters_harmony",
#   bias = c("TSSEnrichment", "log10(nFrags)"),
#   testMethod = "wilcoxon"
# )

# fix incompatible dimensions error (change factor back to logical)
proj_hyp3$DAEG_UP_24_newThr <- as.logical(proj_hyp3$DAEG_UP_24_newThr)

# get marker peaks for DAEG_UP_24_newThr groups
markersPeaks_DAEG_UP_24_newThr <- getMarkerFeatures(
  ArchRProj = proj_hyp3,
  useMatrix = "PeakMatrix",
  groupBy = "DAEG_UP_24_newThr",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  testMethod = "wilcoxon"
)

# get marker peaks for PIMO_pos groups
markersPeaks_PIMO_pos <- getMarkerFeatures(
  ArchRProj = proj_hyp3,
  useMatrix = "PeakMatrix",
  groupBy = "PIMO_pos",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  testMethod = "wilcoxon"
)
```
Took 3 mins. 200 mb matrices, summarized expt object

Failed for markersPeaks_PIMO_pos
> Error in cor(estbgdP, obsbgdP) : incompatible dimensions
maybe need to create Pseudobulk for this PIMO_pos group, and then convert to logical

## Explore Marker Peaks
```{r explore marker peaks}
# SummarizedExperiment containing the marker peaks for DAEG_UP_24_newThr
markersPeaks_DAEG_UP_24_newThr

# marker peaks for each cluster into a List of DataFrame objects
## filtered for FDR and Log2FC
markersPeaks_DAEG_UP_24_newThr_List <- getMarkers(markersPeaks_DAEG_UP_24_newThr, 
                         cutOff = "FDR <= 0.01 & Log2FC >=1")

# access marker peaks for a specific cluster (or cell type)

markersPeaks_DAEG_UP_24_newThr_List$`TRUE`
markersPeaks_DAEG_UP_24_newThr_List$`FALSE`

```
# Plot Marker Peaks

## MA plot, Volcano plot, and Heatmaps

We can use ArchR built in plotting functions to plot both a MA plot and a volcano plot of the marker peaks. Using the `markerPeaks` SummarizedExperiment object we generated above. Here we use `plotMarkers()`, which is a new version of the previous plotting function `markerPlot()`.

```{r plot MA and volcano}

markersPeaks <- markersPeaks_DAEG_UP_24_newThr
# MA plot (at C1 cluster)

pma <- plotMarkers(
  seMarker = markersPeaks, # summarizedExperiment
  name = "TRUE",
  cutOff = "FDR <= 0.1 & abs(Log2FC) >= 1",
  plotAs = "MA" # select MA plot
)

 
# volcano plot (at C1 cluster)
vma <- plotMarkers(
  seMarker = markersPeaks, # summarizedExperiment
  name = "TRUE",
  cutOff = "FDR <= 0.1 & abs(Log2FC) >= 1",
  plotAs = "Volcano" # select Volcano plot
)

# heatmap (at all Clusters_harmony)
heatmapPeaks <- plotMarkerHeatmap(
  seMarker = markersPeaks,
  cutOff = "FDR <= 0.1 & Log2FC >=0.5",
  plotLog2FC = TRUE,
  transpose = TRUE
)

# plotting
pma
vma
draw(heatmapPeaks,
     heatmap_legend_side = "bot", annotation_legend_side = "bot")
```
## Plot Genomic Region of Marker Peaks

```{r genomic region plot of marker peaks}

# check cellColData
proj3@cellColData
sapply(proj3@cellColData, class)

# convert logical to factor for plotting purposes
proj3$DAEG_UP_24_newThr <- as.factor(proj3$DAEG_UP_24_newThr)
  
# select and filter markerPeaks to plot
## note relaxed threshold
featured_peaks <- getMarkers(markersPeaks,
                             cutOff = "FDR <= 0.1 & Log2FC >= 1",
                             returnGR = TRUE)["C1"]
# all clusters
featured_peaks_all <- getMarkers(markersPeaks,
                             cutOff = "FDR <= 0.1 & Log2FC >= 1",
                             returnGR = TRUE)
# plot genomic regions
## locus: HK2
## cell groups: DAEG_UP_24_newThr
## peaks track: featured_peaks from C1 placeholder; should be hypoxia enriched peaks

gr1 <- plotBrowserTrack(
    ArchRProj = proj3, 
    groupBy = "DAEG_UP_24_newThr", 
    geneSymbol = c("HK2"),
    features =  featured_peaks,
    upstream = 50000,
    downstream = 50000,
    title = "HK2 in DAEG_UP_24_newThr cells with \nC1 marker peaks"
)
## locus: HK2
## cell groups: PIMO_pos
## peaks track: featured_peaks from C1 placeholder; should be hypoxia enriched peaks

gr2 <- plotBrowserTrack(
    ArchRProj = proj3, 
    groupBy = "PIMO_pos", 
    geneSymbol = c("HK2"),
    features =  featured_peaks,
    upstream = 50000,
    downstream = 50000,
    title = "HK2 in PIMO_pos cells with \nC1 marker peaks"
)

## show more features
gr3 <- plotBrowserTrack(
    ArchRProj = proj3, 
    groupBy = "PIMO_pos", 
    geneSymbol = c("HK2"),
    features =  featured_peaks_all,
    upstream = 50000,
    downstream = 50000,
    plotSummary = c("bulkTrack", "featureTrack", "geneTrack"), # tracks to show
    size = c(5, 10, 1), # sizes of each plotSummary track
    title = "HK2 in PIMO_pos cells with all Cluster_harmony marker peaks"
)

# show plot
## here we compare cells labelled as DAEG_UP_24_newThr, and those as PIMO_pos at HK2 locus
grid::grid.newpage()
grid::grid.draw(gr1$HK2)
grid::grid.newpage()
grid::grid.draw(gr2$HK2)
grid::grid.newpage()
grid::grid.draw(gr3$HK2)
grid::grid.draw(gr4$HK2)
```

Here when I plotted all the marker peaks (peaks for each Harmony Clusters), the track was squished due to the high number of clusters to plot. I added the `plotSummary` and `sizes` parameter to adjust the track sizes to limit the squishing.

# Pairwise Testing between groups

ArchR also allows the pairwise comparison within a selected group from `cellColData`, such as `Clusters` or `Harmony_clusters` and choose a pair of subgroups within that larger group to compare. For example, we can compare C1 vs C2, or if we have cell types, `Erythroid` cells vs `Progenitor` cells. 

Another example of a group in `cellColData` is the `PIMO_pos` column, which can be selected by `groupBy` and then we can compare the cells that were assigned by AUCell as `TRUE` vs `FALSE`. We can set `useGroups = "TRUE"` and `bdgGroups = "FALSE"`, which allows us to find peaks enriched in hypoxic cells.

```{r pairwise, eval = FALSE}
proj_hyp3$DAEG_UP_24_newThr <- as.logical(proj_hyp3$DAEG_UP_24_newThr)

markerTest <- getMarkerFeatures(
  ArchRProj = proj_hyp3,
  useMatrix = "PeakMatrix",
  groupBy = "DAEG_UP_24_newThr",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  useGroups = "TRUE",
  bgdGroups = "FALSE"
)
```

> Error in cor(estbgdP, obsbgdP) : incompatible dimensions
Potential error due to lack of pseudobulk of PIMO_pos cells. Note, the proj3 object pseudobulk is run on the `Harmony_clusters`.

```{r plot pairwise results}
pwMA <- markerPlot(seMarker = markerTest, name = "TRUE", 
                   cutOff = "FDR <= 0.1 & abs(Log2FC) >= 1", plotAs = "MA")
pwVolc <- markerPlot(seMarker = markerTest, name = "TRUE", 
                   cutOff = "FDR <= 0.1 & abs(Log2FC) >= 1", plotAs = "Volcano")


pwMA
pwVolc


table(proj_hyp3$PIMO_pos)
table(proj_hyp3$DAEG_UP_24)
table(proj_hyp3$DAEG_UP_24_newThr)
```

# Motif Enrichment

## Saving and Loading an ArchRProject
```{r saving and loading project, eval = FALSE}
# check where output directory is assigned
getOutputDirectory(proj) # Terekhanova/

# save ArchRProject object and associated arrowfiles to output directory
proj <- saveArchRProject(ArchRProj = proj,
                         outputDirectory = "Save-proj")

# proj_hyp is proj with pseudobulk via group coverages
proj_hyp <- saveArchRProject(ArchRProj = proj_hyp,
                             outputDirectory = "Save-proj_hyp")

# proj_hyp2 has pseudobulk and macs2 peaks
proj_hyp2 <- saveArchRProject(ArchRProj = proj_hyp2
                              outputDirectory = "Save-proj_hyp2")
# load
proj <- loadArchRProject(path = "Terekhanova")
#proj2 <- loadArchRProject(path = "Terekhanova")
```