---
title: "ArchR_tutorial"
author: "Ronald Wu"
date: '2024-05-03'
output: html_document
---

# Single Cell ATAC-seq Analysis
There are a few well known packages in R and python that can be used to analyze scATAC-seq data.
+ R Packages:
  + Signac: developed by Tim Stuart while in the Satija lab. Actively maintained by Stuart lab.
    + [website](https://stuartlab.org/signac/)
  + ArchR: 
    + [website](https://www.archrproject.com/)

# ArchR Vignettes
The purpose of this notebook is to learn how to use `ArchR` for processing and analyzing scATAC-seq data. We'll be following the vignettes provided by the Greenleaf lab, which is the original developer of this package. All code is slightly modified from their original vignettes from the following:
+ [brief tutorial](https://www.archrproject.com/articles/Articles/tutorial.html)
+ 

# Setup
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Install packages

```{r install packages, eval = FALSE}
# install devtools: allows install of packages from github
if (!requireNamespace("devtools", quietly = TRUE)) install.packages("devtools")

# if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

devtools::install_github("GreenleafLab/ArchR", ref="master", repos = BiocManager::repositories())

library(ArchR)
ArchR::installExtraPackages() # installs Cairo and other stuff

 library(BiocManager)
      install("BSgenome.Hsapiens.UCSC.hg19") # needed to set addArchRGenome to hg19

install.packages("hexbin") # for plotting
```

I also installed `macs2` version 2.2.9.1 locally (on macbook pro) using the command `pip install macs2`. MACS2 will be called by ArchR for peak calling purposes. Note: There is a newer versions of MACS called `MACS3` but I don't believe is it support by ArchR.

# Load Libraries
```{r load libraries}
library(ArchR) # version 1.0.2
```

# 1. Brief tutorial

## Setup and download tutorial data
Below we are following the [brief tutorial](https://www.archrproject.com/articles/Articles/tutorial.html).

```{r Setup and download tutorial data, eval = FALSE}
library(ArchR)
set.seed(1)
addArchRThreads(threads = 2) 

# download tutorial data (0.5 gigs to current working directory)
inputFiles <- getTutorialData("Hematopoiesis")

# change location of tutorial files to data/
temp <- names(inputFiles)
inputFiles <- paste0("data/", inputFiles)
names(inputFiles) <- temp
```
After downloading the tutorial files, we manually moved them to the `data/` folder in the current working directory.

## Creating Arrow Files

```{r creating arrow files}
# set genome to hg19
addArchRGenome("hg19")

# create arrow files
ArrowFiles <- createArrowFiles(
  inputFiles = inputFiles[1],
  sampleNames = names(inputFiles)[1],
  minTSS = 4, # Dont set this too high because you can always increase later
  minFrags = 1000, 
  addTileMat = TRUE,
  addGeneScoreMat = TRUE
)

ArrowFiles
```
Note: we modified the above tutorial code to process only 1 of 3 samples. Had an error when all 3 inputFiles were used.

When only the first sample `scATAC_BMMC_R1` was used, arrow file was created without error. Still took 32 mins to complete.

Two matrices are created: 
1) GeneScoreMatrix: the atac-seq signal proximal to the TSS is used to make an estimate of the gene activity score.
2) TileMatrix: A counts matrix where the genome is tiled into a fixed-width sliding window. Not the usual peak based matrix.

## Inferring doublets

```{r inferring doublets}
doubScores <- addDoubletScores(
  input = ArrowFiles,
  k = 10, #Refers to how many cells near a "pseudo-doublet" to count.
  knnMethod = "UMAP", #Refers to the embedding to use for nearest neighbor search.
  LSIMethod = 1
)

doubScores
```

The doublet results are found in the `QualityControl/scATAC_BMMC_R1` folder.

## Creating an ArchRProject

```{r create ArchRProject}
proj <- ArchRProject(
  ArrowFiles = ArrowFiles, 
  outputDirectory = "HemeTutorial",
  copyArrows = TRUE #This is recommended so that you maintain an unaltered copy for later usage.
)

# show available matrices for the project 
getAvailableMatrices(proj) #  "GeneScoreMatrix" "TileMatrix" 

# filter doublets; note addDoubletScores must be run previously

proj <- filterDoublets(ArchRProj = proj)
```

An ArchRProject must be created for further analysis of our arrowfiles. About 4.9% cells were doublets and filtered out. Thus, 4689 cells were retained.

## Dimensionality Reduction and Clustering

```{r dimensionality reduction and cluster}
# IterativeLST used for dimensionality reduction
proj <- addIterativeLSI(ArchRProj = proj, useMatrix = "TileMatrix", name = "IterativeLSI")

# runs seurat's FindClusters function
proj <- addClusters(input = proj, 
                    reducedDims = "IterativeLSI" # name of the reducedDims object
                    )


```

IterativeLSI is used for dimensionality reduction. By default, the log(tf - idf) method is used. Can also be changed to tf - log(idf) or log(tf) - log(idf).

There are 8 Clusters identified.

## Visualizing in a 2D UMAP Embedding

```{r UMAP Visualization}
# add UMAP embedding
proj <- addUMAP(ArchRProj = proj, reducedDims = "IterativeLSI")

# plot UMAP:
## colorBy Sample
p1 <- plotEmbedding(ArchRProj = proj, colorBy = "cellColData", name = "Sample", embedding = "UMAP")
## colorBy Clusters
p2 <- plotEmbedding(ArchRProj = proj, colorBy = "cellColData", name = "Clusters", embedding = "UMAP")

p1 + p2

# ggAlignPlots(p1, p2, type = "h")
```
### save plots to pdf
```{r save plots to pdf, eval = FALSE}
plotPDF(p1,p2, # vector of plots to plot
        name = "2024-05-06-BMMC-Plot-UMAP-Sample-Clusters.pdf", # file name
        ArchRProj = proj, 
        addDOC = FALSE, width = 5, height = 5)
```

## Assigning Clusters with Gene Scores

Given a vector of marker genes, we can overlay the gene scores of those markers onto the UMAP.

```{r cluster assignment with gene scores}

# smooth dropout noise in gene scores
proj <- addImputeWeights(proj)

# marker genes to use to assign clusters
markerGenes  <- c(
    "CD34",  #Early Progenitor
    "GATA1", #Erythroid
    "PAX5", "MS4A1", "MME", #B-Cell Trajectory
    "CD14", "MPO", #Monocytes
    "CD3D", "CD8A"#TCells
  )
```

```{r plot umap with marker genes}
# plot UMAP 
## colorBy markerGenes 
p <- plotEmbedding(
    ArchRProj = proj, 
    colorBy = "GeneScoreMatrix", 
    name = markerGenes, 
    embedding = "UMAP",
    imputeWeights = getImputeWeights(proj)
)

# plot a specific marker
p$CD14

#Rearrange for grid plotting
p3 <- lapply(p, function(x){
    x + guides(color = FALSE, fill = FALSE) + 
    theme_ArchR(baseSize = 6.5) +
    theme(plot.margin = unit(c(0, 0, 0, 0), "cm")) +
    theme(
        axis.text.x=element_blank(), 
        axis.ticks.x=element_blank(), 
        axis.text.y=element_blank(), 
        axis.ticks.y=element_blank()
    )
})
do.call(cowplot::plot_grid, c(list(ncol = 3),p3))
```

The MAGIC diffusion matrix is used smooth dropout noise in the gene score matrix. 

### save vector of plots to pdf

```{r save vector of plots to pdf}
plotPDF(plotList = p, 
    name = "2024-05-06-BMMC-Plot-UMAP-Marker-Genes-W-Imputation.pdf", 
    ArchRProj = proj, # plots are saved into Plots/ in the assigned outputDirectory
    addDOC = FALSE, width = 5, height = 5)
```

## Visualizing Genome Browser Tracks
```{r Visualizing Genome Browser Tracks}

# visualize the genome by the marker genes and by clusters
p4 <- plotBrowserTrack(
    ArchRProj = proj, 
    groupBy = "Clusters", 
    geneSymbol = markerGenes, 
    upstream = 50000,
    downstream = 50000
)

# plot a specific gene
grid::grid.newpage()
grid::grid.draw(p4$CD14)
grid::grid.draw(p4$GATA1)
```
### save PDF of genome browser tracks
```{r pdf of genome browser tracks}

plotPDF(plotList = p4, 
    name = "2024-05-06-Plot-Tracks-Marker-Genes.pdf", 
    ArchRProj = proj, 
    addDOC = FALSE, width = 5, height = 5)
```

## Saving and Loading an ArchRProject
```{r saving and loading project}
# save
# proj <- saveArchRProject(ArchRProj = proj)

# load
# proj <- loadArchRProject(path = "HemeTutorial")
```

By default, saving the project will overwrite the files in the output directory `HemeTutorial/`, which we assigned earlier when creating the ArchRProject object.

## Session info
```{r session info}
Sys.Date()
sessionInfo()
```




